### ISSUE TITLE
RFC: Rust everywhere

### ISSUE THREAD

**Issue by gkelly** [lowRISC/opentitan#1225]
This isn't an RFC-Process-RFC, this is a testing-the-waters-RFC.  Last week @sriyerg mentioned that he'd be OK writing in Rust.  That begs the question: why don't we just write everything in rust?  The boot stack, etc.  We're already going to be using Tock, with userspace applications written in Rust, so why stop there?

What would this entail?
  * Switching our existing software tree over to Rust.
    * Figuring out what a DIF under Rust looks like (I don't anticipate there's much to change here.)
    * Porting our in-progress DIFs to Rust.
  * (Probably) replacing our build system with Cargo.

Pros:
  * Unifies the language base for the whole project.
  * Moves us over to cargo for the build system, which has very clear semantics for building the code and dependency management.
  * It's a strongly-typed memory-safe language that eliminates whole classes of bugs when used properly (this is more subtle when writing low-level code which is going to be sprinkled with `unsafe` blocks).

Cons:
  * A new language: new compiler, new bugs, etc.
  * Very low familiarity with rust among existing contributors.
  * Relatively unproven in the DV and mask ROM world.

This issue is a temperature check.  If you have opinions in either direction please contribute them here.  The outcome of this discussion will inform whether or not we move this to a true RFC and attempt to get it ratified by the TC.

**Comment by mcy** [comment]
I'd like to point out: Rust has generics, which are a huge pro for us. A lot of the work we do relies on late-binding of fundamental functions (such as device properties), and generics provide us with a robust mechanism for doing so.

Rust also means we get to throw out all of my foundational libbase work, which simplifies our environment significantly.

Also, I keep a close pulse on Rust compiler forums, so I am available for Rust language-lawyering (which Sam and I already do for C), though I have little experience inside rustc's codebase proper. It uses LLVM as a backend, but it has different aliasing rules and, as such, a different optimizer profile.

Two cons Garret doesn't mention:
- We probably need to use intrinsics and inline assembly, which are nightly-compiler only. This is already an issue with Tock though, so I don't think we have to worry about it.
- Rust's inline assembly syntax is an unstable knockoff of Clang's syntax. There are active talks to replace it with a stable syntax that matches Rust's other features. I am closely involved with the proposal, to ensure that it has well-documented semantics, and RISC-V support from day 1.

**Comment by gkelly** [comment]
> I'd like to point out: Rust has generics, which are a huge pro for us. A lot of the work we do relies on late-binding of fundamental functions (such as device properties), and generics provide us with a robust mechanism for doing so.
> 
> Rust also means we get to throw out all of my foundational libbase work, which simplifies our environment significantly.
> 
> Also, I keep a close pulse on Rust compiler forums, so I am available for Rust language-lawyering (which Sam and I already do for C), though I have little experience inside rustc's codebase proper. It uses LLVM as a backend, but it has different aliasing rules and, as such, a different optimizer profile.

We have LLVM expertise/developers through people at lowRISC that are directly involved in the project, which is another pro I think.

> Two cons Garret doesn't mention:
> 
> * We probably need to use intrinsics and inline assembly, which are nightly-compiler only. This is already an issue with Tock though, so I don't think we have to worry about it.
> * Rust's inline assembly syntax is an unstable knockoff of Clang's syntax. There are active talks to replace it with a stable syntax that matches Rust's other features. I am closely involved with the proposal, to ensure that it has well-documented semantics, and RISC-V support from day 1.

These are good points, thanks!

**Comment by tomoveu** [comment]
Based on my latest interaction with the person who is driving Bare-metal Rust for ST Microelectronics STM32, deploying native firmware in Rust is not something ready for production. Please correct me if I am missing new information. 

Last I checked, there is still a heavy usage of bindings bewteen C BSP and libraries before we get to actually writting in Rust the firmware. At least for the critical sections like boot process I would prefer to keep it small and clean, rather than mix of ASM, C and Rust. 

My feedback is based on the progress seen for STM32.

**Comment by gkelly** [comment]
> Based on my latest interaction with the person who is driving Bare-metal Rust for ST Microelectronics STM32, deploying native firmware in Rust is not something ready for production. Please correct me if I am missing new information.
> 
> Last I checked, there is still a heavy usage of bindings bewteen C BSP and libraries before we get to actually writting in Rust the firmware. At least for the critical sections of OpenTitan like boot process I would prefer to keep it small and clean, rather than mix of ASM, C and Rust.

I believe that use is dominated by links into stm32cube code, which is an extensive peripheral support library for STM parts.  One of the benefits of OpenTitan is that we're not hindered by an extensive peripheral support library codebase. :wink:

Yes, there is some assembly needed for getting out of reset.  I have seen this done as inline assembly in naked linkage fns in rust, but I think the current state of the art is just to write the ~75 instructions and link them statically.  I'm unsure if this style is motivated by the instability in inline `asm` in rust or if it's that once you write those early instructions correctly you rarely (if ever) need to change them.

> My feedback is based on the progress seen for STM32.

Thanks for sharing your feedback, @tomoveu!

**Comment by sriyerg** [comment]
I did say that I *personally* was interested in learning to code in Rust. If we should really use it in favor of C as the language of choice for OpenTitan DV work needs a lot more discussion, in relation to these points: 

1. All DV engineers already are familiar with C and can start contributing from day 1
2. If we go with Rust then **all** DV engineers who sign up to work on SoC level DV need to be trained, not just me
  - Note that even the C skills most DV engineers possess are rusty (pun intended) at best, so having to learn Rust *may be* harder 
  - How will this work if we assign new DV folks in the middle of the project? Can we setup an on-demand Rust training?
  - Situations where we may need to move resources around to meet our schedule will end up being problematic
3. DV schedules are considerably more aggressive than SW (need to signoff well ahead of TO) - the overhead of having to learn a whole new language will very likely blow up our schedule 
4. Our only interest is verifying that the HW is behaving correctly so we are perfectly happy with very simple bare-metal C tests in our DV testbench (occasionally hand-shaking with the SystemVerilog world). The IPs that make up the SoC are already being thoroughly verified at their levels, so the scope of finding bugs at the top level is low. Plus, we are not building super complex OS / software stack - so the pros that Rust offers over C are possibly not very relevant for DV work 
5. There are other embedded C <--> DV testbench development efforts from DV teams working on other projects (at Google) with whom we will no longer be able to collaborate / share ideas with if we move to a different language paradigm
6. In terms of an acquired skill career development-wise, it doesn't appear to be a high value proposition either, given that the chances of finding DV jobs in the ASIC industry that also require an experience in Rust are slim to none.  

Having thought through these, I am leaning towards sticking with C for DV work. I am interested in hearing from Si / project leads specifically on how they feel about this overhead affecting our project schedule. 
 
-Sri

**Comment by tomoveu** [comment]
@gkelly I can confirm that without the BPS and Peripheral library dependencies, it is a matter of one ASM startup file, but there is one detail.

- OpenTitan would need a BSP as well, with or without peripherals. 
- Low-level initialization Rust code requires a lot of use of the `unsafe ` keyword
- That eliminates the memory safe benefit for the most critical section, the system boot

The benefit of Rust definitely comes at the next stage - User application. OpenTitan is initialized and from there we execute our arbitrary code in `safe `Rust.

@sriyerg rises several valid concerns about adoption and contributions. I am not considering the timeline when saying this, but focusing on the first two aspects - "adoption and contributions".

I think having a minimalistic C boot code designed with Rust application in mind would be a good approach. This way we keep the possibility open for everyone to collaborate.

I must agree that doing innovation in Hardware and Software at the same time is tempting, but I do not know if the project structure allows to split the focus between both. I am interested in all-rust solution and I would be happy to help with rust boot code as an iteration after the C boot code version. This way we first complete the Hardware innovation and then move on to complete the Software innovation :)

Just my 2 cents.

**Comment by sjgitty** [comment]
> I am interested in hearing from Si / project leads specifically on how they feel about this overhead affecting our project schedule.

Thanks a lot for the detailed thoughts, @sriyerg. All good points. 3) is the one I'm most acutely aware of; I've never been on a chip project where DV wasn't the long pole and our current DV personnel balance is not ideal. 5) *might* not be an issue, depending upon what the Rust-DPI solution looks like, which its own issue that you haven't highlighted here.

I'm wondering if there isn't a half-measure here. We're heading down a path where we're building two pillars upon a single foundation of DIFs: the software stack for boot loaders, OS and applications; and the hardware stack for verification and validation. By my estimation, the DIFs will comprise a smallish percentage of the total SW content. We've highlighted it as one of the most critical components: worthy of its own style guide (TODO); worthy of what we HWEs call "SWE-level scrutiny"; hopefully portable to all of these platforms; and hopefully productized not just as reference code (TBD).

Is it possible - just testing the theory, at this point - to have Rust for DIFs, enabling Rust (and Tock) for the other SW collateral, but allowing all HW (DV at a minimum) work layered on top of it to be in C? This would alleviate a lot of Sri's concerns. What would the environment look like? Is the Rust <=> C interaction unidirectional, or can it be layered like this? For that matter, if it can't, we would half to quantify the effort of any Rust <=> SystemVerilog DPI, which I think might be its own dealbreaker. (Otherwise I would have assumed you could do Rust <=> C <=> DPI : SystemVerilog. I searched for a Rust <=> SystemVerilog DPI and came up empty.)

Can anyone answer those points?

**Comment by mcy** [comment]
I want to answer a few themantic concerns, in no particular order:

---

Getting out of reset, in the current design, is almost no effort:
```S
.globl _start
// Optional register clobber.
li x1, 0x0
// ...
li x31, 0x0

la sp, _stack_start
call _rust_start
// |_rust_start()| is, of course, responsible for setting up .bss, .data, and so on.
```
We still need to set up the IRQ vector but that's just linker shenanigans at that point, which we cannot avoid.

---

I admit that teaching Rust isn't easy, but the subset that DV needs is, essentially, a reskinned C with none of the stupid baggage. To dispel preconceptions:
- Rust is not a functional programming language. It is imperative, like C.
- Rust produces the same artifacts a C compiler does; it even shares a backend (LLVM) with `clang`.
- Rust has a lot of advanced features, like C++. However, in an embedded environment where the most common types are "word" and "pointer to word", the language can be reduced substantially.

Things you need to learn, at an absolute minimum:
- Control flow (you get ifs and while loops out of the box, which is probably all you need).
- Pointer manipulation (you'll be using raw pointers, so you'll write `ptr.read()` instead of `*ptr`).
- It's unlikely that DV (I assume) needs to device complex, non-plain-old-data structs, so no need to mess with traits, impls, or any of that.

Things that you get in Rust that you don't in C:
- Namespacing that isn't absolutely insane.
- Undefined behavior rules that don't require me or Sam to read between the lines of the ISO standard.
- MAYBE: A more expressive language that SW can use to build abstractions that could help DV be more productive in their specific tasks.

If we decide to go this route, I can probably concoct a short Rust class that teaches what DV needs to be productive without getting into the details you don't care about (DV does not have any benefit to gain from borrow checking, IMO).

There's also a large effort to build tools for mechanically translating C source code into Rust. https://c2rust.com/ is one of many examples. The output isn't amazing (mostly due to C's baggage, like its integer types) but I think it shows that using Rust as if it were C is a realistic possibility.

---

> Is it possible - just testing the theory, at this point - to have Rust for DIFs, enabling Rust (and Tock) for the other SW collateral, but allowing all HW (DV at a minimum) work layered on top of it to be in C?

Yes! This is a problem I've been thinking about a lot. Rust has amazing C interop, because it has no weird runtime that would require something more complicated, and, ultimately, `rustc` produces object files very similar to those `clang` produces (almost as if they use the same backend... =P).

I can write the following rust:
```rust
#[no_mangle]
extern "C" dv_bridge_do_thing(x: u32) -> u32 { ... }
```
and the following C:
```
uint32_t dv_bridge_do_thing(uint32_t);
// ...
dv_bridge_do_thing(kMyConstant);
```

The `#[no_mangle]` annotation turns off name mangling, so C and Rust see the same name, and `extern "C"` makes the function have the C ABI, instead of Rust's more complicated calling convention. This way, we can build a module of bridge functions that allow C code to call into Rust. It's completely possible to, for example, implement libc as a combination of `.h` files containing libc definitions, and `.rs` files containing Rust implementations of the C symbols.

While the "official" DIFs will all use the Rust ABI (aka `extern "Rust"`), it would be very easy to write a module of bridge functions consisting of `.h` files with definitions for C, and `.rs` files full of `extern "C"` functions that call the official DIFs. Inlining and LTO will probably close the performance gap somewhat, too. C++'s interop with C is very similar in this regard: `extern "C" {}` in C++ is the inspiration for the same feature in Rust.

I have not investigated this, but I have to wonder if tools exist to do this automatically... if they don't, it would be very easy to build them, though. There are excellent libraries out there for parsing and emitting Rust and C source code.

Also, there is no limit on layering here: Rust function calls are just function calls, and, especially in an embedded environment with a minimal runtime, SV calling C calling Rust calling C calling Rust calling ... is absolutely supported.

---

> worthy of what we HWEs call "SWE-level scrutiny"

😁

**Comment by thedataking** [comment]
> There's also a large effort to build tools for mechanically translating C source code into Rust. https://c2rust.com/ is one of many examples. The output isn't amazing (mostly due to C's baggage, like its integer types) but I think it shows that using Rust as if it were C is a realistic possibility.

C2Rust contributor here. 100% agree that the output isn't amazing. I do want to point out that after we got the C2Rust transpiler into decent shape, we have focused more about refactoring and cleaning up the output. We blogged about improvements around header files here: https://immunant.com/blog/2019/12/header_merging/. We're also going to blog about automatic cleanup of integer types and casts. We are also able to convert raw pointers into Rust-native references although much work remains there. This is all to say that C2Rust is a moving target and we're slowly making the output less awful/more awesome & safe 😁

Very happy to support this use case. We're also seeing interest from other Googlers and gave a Tech talk on the subject on October 23rd of last year.
